java/java.j2seplatform/src/org/netbeans/modules/java/j2seplatform/JDKDetectorUtils.java

```diff
@@ -0,0 +1,81 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.j2seplatform;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.netbeans.api.java.platform.JavaPlatform;
+import org.netbeans.api.java.platform.JavaPlatformManager;
+import org.netbeans.modules.java.j2seplatform.platformdefinition.J2SEPlatformImpl;
+import org.netbeans.spi.java.platform.JavaPlatformFactory;
+import org.openide.filesystems.FileUtil;
+import org.openide.util.Lookup;
+
+/**
+ *
+ * @author mbien
+ */
+class JDKDetectorUtils {
+
+    /**
+     * Registers new JDKs if they aren't registered yet.
+     * @param jdks List of paths to the JDKs.
+     * @param namer Converts a Path to the display name of the JDK.
+     */
+    static void registerJDKs(List<Path> jdks, Function<Path, String> namer) {
+
+        List<JavaPlatform> platforms = Stream.of(JavaPlatformManager.getDefault().getInstalledPlatforms())
+                                             .filter(JavaPlatform::isValid)
+                                             .collect(Collectors.toList());
+
+        Set<String> registeredJDKNames = platforms.stream()
+                                                  .map(JavaPlatform::getDisplayName)
+                                                  .collect(Collectors.toSet());
+
+        Set<Path> registeredPaths = platforms.stream()
+                                             .flatMap(p -> p.getInstallFolders().stream())
+                                             .map(f -> Paths.get(f.getPath()))
+                                             .collect(Collectors.toSet());
+
+        for (JavaPlatformFactory.Provider provider : Lookup.getDefault().lookupAll(JavaPlatformFactory.Provider.class)) {
+
+            JavaPlatformFactory platformFactory = provider.forType(J2SEPlatformImpl.PLATFORM_J2SE);
+            if (platformFactory != null) {
+                for (Path jdk : jdks) {
+                    try {
+                        String jdkName = namer.apply(jdk);
+                        // don't register if something with the same name or same path exists
```

Comment:
The `namer` should really only be applied if that path doesn't already have the file attribute for the name?

We could do with a better way of linking path and chosen name - the idea was to allow re-registered JDK's to get the previously set user name by default.

It's probably a minor concern now you're checking for path rather than name to determine registration, but could still be worth considering?

Reply:
> The namer should really only be applied if that path doesn't already have the file attribute for the name? Otherwise it should be the value of that attribute.

my thinking was to not mess with already registered JDKs at all if there is a conflict in name or path. If the user changed something, the user should remove it first before the automation kicks in again. Being too smart here by updating things quietly during restarts could cause issues.

Keep in mind the version before that simply registered JDKs without any checks :)

Reply:
actually you are probably right. What if SDKMAN is updating JDKs in place. The old version would have simply overwritten the registered platforms. This PR would keep the old ones until they are manually removed.

probably better to think about this more

Reply:
SDK man does not update JDK in place as far as I know. It downloads the new one and sets the current symlink to the one used.

Reply:
@neilcsmith-net i tested this:
```java
                        String jdkName = null;
                        FileObject fo = FileUtil.toFileObject(jdk);
                        if (fo != null) {
                            Object value = fo.getAttribute("J2SEPlatform.displayName"); // NOI18N
                            if (value instanceof String) {
                                jdkName = (String) value;
                            }
                        }
                        if (jdkName == null) {
                            jdkName = namer.apply(jdk);
                        }
```
and I believe this is worse since it can lead to very weird behavior. If you add a JDK with a custom name, then remove it again, it will be added with the same name again on next start which looks for the user as if removal didn't work.

Having automated JDK detection use a fixed naming pattern is better IMO, since it automatically communicates where this is from.

@lkishalmi thanks for checking!

Reply:
@mbien yes, that sounds wrong. One possible thought would be to _ignore_ folders with that attribute? That should stop removed folders from being automatically re-added?

Otherwise, leave as is for now. The naming / removal history could probably do with some tweaking at a later date. In retrospect the file attribute was not the right way to do this anyway.

Reply:
interesting idea.
```java
                    FileObject fo = FileUtil.toFileObject(jdk);
                    if (fo != null) {
                        Object value = fo.getAttribute("J2SEPlatform.displayName"); // NOI18N
                        if (value instanceof String) {
                            break; // skip, this indicates that it was added (and potentially removed) by the user
                        }
                    }
```
i think this has potential but there is one big issue: there is no reset button. We would have to be prepared to tell users to remove nb_userdir/var/attributes.xml

Reply:
however :)

having to remove this check later is probably the better option than not having it -> going to change it

Reply:
Why is the lack of a reset button an issue? If a JDK registration has been manually removed, shouldn't it need to be manually re-added?

That attribute was added primarily to try and maintain the names given to Disco downloaded JDKs incidentally. So they could be removed and then re-added under same name. I've had doubts since about that way to do it, but somewhere I think we should store metadata like names, removal status, etc. of manually unregistered JDKs.

Reply:
> Why is the lack of a reset button an issue? If a JDK registration has been manually removed, shouldn't it need to be manually re-added?

the longer I thought about it the more I liked the idea of skipping those folders. I think it is the safest option to get auto detection in.

step 1: don't ever override already registered JDKs
step 2: don't annoy the user by registering stuff the user potentially just removed

having a growing "block list" without reset button is intuitively wrong to me, but it is probably completely fine in this particular case - and the better option to not having it.

Reply:
It'll also reset on IDE upgrade - those attributes don't get imported (one of the issues with using them for this IMO). I think it would be good to get this in for NB18, then possibly rethink how we store info in NB19, and whether we need a reset.